src/rgw/driver/rados/rgw_pubsub_push.cc

```diff
@@ -131,57 +132,6 @@ class RGWPubSubHTTPEndpoint : public RGWPubSubEndpoint {
   }
 };
 
-namespace {
-// this allows waiting untill "finish()" is called from a different thread
-// waiting could be blocking the waiting thread or yielding, depending
-// with compilation flag support and whether the optional_yield is set
-class Waiter {
-  using Signature = void(boost::system::error_code);
-  using Completion = ceph::async::Completion<Signature>;
-  std::unique_ptr<Completion> completion = nullptr;
-  int ret;
-
-  bool done = false;
-  mutable std::mutex lock;
-  mutable std::condition_variable cond;
-
-public:
-  int wait(const DoutPrefixProvider* dpp, optional_yield y) {
-    std::unique_lock l{lock};
-    if (done) {
-      return ret;
-    }
-    if (y) {
-      boost::system::error_code ec;
-      auto yield = y.get_yield_context();
-      auto&& token = yield[ec];
-      boost::asio::async_initiate<boost::asio::yield_context, Signature>(
-          [this, &l] (auto handler, auto ex) {
-            completion = Completion::create(ex, std::move(handler));
-            l.unlock(); // unlock before suspend
-          }, token, yield.get_executor());
-      return -ec.value();
-    }
-    maybe_warn_about_blocking(dpp);
-
-    cond.wait(l, [this]{return (done==true);});
```

Comment:
the existing Waiter handles both states of optional_yield. ceph::async::yield_waiter only accepts the yield_context, and calling y.get_yield_context() on a null_yield is undefined behavior 

i think this Waiter should use ceph::async::yield_waiter internally, while preserving the mutex/condition_variable for the null_yield case

Reply:
currently, these functions are called either from the frontend or from the notifications manager.
both are not using the `null_yield`. to avoid undefined behavior, i can add an assert

Reply:
don't we also send notifications from background threads like lifecycle and multisite? those would pass null_yield

Reply:
yes, you are right (forgot about these)

Reply:
yes, you are right (forgot about these)

Reply:
also, no coverage for multisite sync notifications: https://tracker.ceph.com/issues/67178

Reply:
for testing, you can disable `rgw_beast_enable_async` so that frontend requests always see null_yield

Reply:
kafka tests are passing locally with the flag set to "false" (`vstart.sh -n -d -o "rgw_beast_enable_async=false"`)

can we set the flag from commandline in a test, so we can cover that in teuthology as well?